Add auto_linear formulas and some others

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #69956
* #69998
* #69884
* __->__ #69727

Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

soulitzer committed:
Add auto_linear formulas and some others

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"




[ghstack-poisoned]

albanD commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -11251,8 +11267,8 @@ def ref_pairwise_distance(input1, input2):
         ref=lambda x, threshold, value: np.where(x > threshold, x, value).astype(x.dtype),
         dtypes=all_types_and(torch.bfloat16),
         dtypesIfCUDA=all_types_and(torch.float16, torch.bfloat16),
-        supports_forward_ad=False,
```
Do you know why these were specified?

soulitzer replied:
Hmm not sure, maybe the author just wanted to be explicit that forward isn't implemented?

albanD reviewed (commented):
Do we have a plan on how to easily test the `*_backward()` formulas?

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

[ghstack-poisoned]

soulitzer commented:
> Do we have a plan on how to easily test the *_backward() formulas?

Added a new PR in this stack that does this right before this one.

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

[ghstack-poisoned]

soulitzer commented:
We have `4` lines of comments and `46` lines of formulas in `derivatives.yaml`. See that they are tested below and adds up to `46`:
```
4 (tested with OpInfo)
softplus
softshrink
threshold
threshold_
-------------------------------------------
12 (tested with OpInfo)
reflection_pad{1,2,3}d
reflection_pad{1,2,3}d_backward
replication_pad{1,2,3}d
replication{1,2,3}d_backward

Testing:
- Looking at OpInfo for reflect and replicate variants of functional.pad.
- grad and fw-over-rev gradgrad are supported
- 1d, 2d, 3d corresponds to 3d, 4d, 5d
-------------------------------------------
10 (mostly OpInfo)
upsample_linear1d.vec
upsample_linear1d_backward.vec
upsample_bilinear2d{,_aa}.vec (test_nn)
upsample_bilinear2d{,_aa}_backward.vec (test_nn)
upsample_trilinear3d.vec
upsample_trilinear3d_backward.vec
upsample_bicubic2d.vec
upsample_bicubic2d_backward.vec

6 (OpInfo)
upsample_nearest{1,2,3}d.vec
upsample_nearest{1,2,3}d_backward.vec

6 (test_nn)
upsample_nearest_exact{1,2,3}d.vec
upsample_nearest_exact{1,2,3}d_backward.vec

Testing:
- Looking at 'interpolate' OpInfo since upsample is deprecated
- only vec is tested, rest are for bc reasons only
- interpolate support grad and fw-over-rev gradgrad
- 1d, 2d, 3d corresponds to 3d, 4d, 5d
- aa is different, added manual tests for forward and backward
-------------------------------------------
8
{,_adaptive_}avg_pool{2,3}d
{,_adaptive_}avg_pool{2,3}d_backward

OpInfos: 
- nn.functional.{,adaptive}_avg_pool{1,2,3}d (grad, gradgrad)
- each support grad and fw-over-rev gradgrad
-------------------------------------------
total 46
-------------------------------------------
```

soulitzer commented:
Update: made sure that all formulas added are tested (see comment abovee). Because non-vec upsample overloads are only kept for backward compatibility (and not tested) I'm not adding forward AD formulas for those.

albanD reviewed (commented):
Thanks for the details on the tests. That looks good.

Some CI and var division by 0 and that will be good to go I think.

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33031728).

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

albanD reviewed (commented):
The change looks good but the xla failure is real. We need to prepare a sister PR for this one.

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer commented:
Since forward AD doesn't actually work with XLA, perhaps we should just skip actually in this PR.

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

albanD reviewed (approved):
SGTM

Lint failure is unrelated right?

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33031728).

soulitzer committed:
Update on "Add auto_linear formulas and some others"


Still need to test the backward ones. We would need to update gradgradcheck to check forward over backward.

Differential Revision: [D33031728](https://our.internmc.facebook.com/intern/diff/D33031728)

[ghstack-poisoned]

soulitzer commented:
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33031728).

soulitzer commented:
> Lint failure is unrelated right?

Yeah I think so, the failures should go away in this latest rebase
